core/src/main/java/org/elasticsearch/common/rounding/TimeZoneRounding.java

```diff
@@ -218,7 +219,19 @@ public byte id() {
         public long roundKey(long utcMillis) {
             long timeLocal = timeZone.convertUTCToLocal(utcMillis);
             long rounded = Rounding.Interval.roundValue(Rounding.Interval.roundKey(timeLocal, interval), interval);
-            return timeZone.convertLocalToUTC(rounded, false, utcMillis);
+            try {
+                return timeZone.convertLocalToUTC(rounded, true, utcMillis);
+            } catch (IllegalInstantException e) {
+                /*
+                 * The rounded local time is illegal and landed in a DST gap. In
+                 * this case, we choose 1ms tick after the transition date. We
+                 * don't want the transition date itself because those dates,
+                 * when rounded themselves, fall into the previous interval.
+                 * This would violate the invariant that the rounding operation
+                 * should be idempotent.
+                 */
+                return timeZone.previousTransition(utcMillis) + 1;
```

Comment:
We should document the fact that it will creates buckets smaller than the interval ? Only when using timezone though so we should also add a line about the limitations of using a timezone conversion.